### DISCUSSION TITLE
A more cautious approach to securing transactions

### DISCUSSION THREAD

**Discussion by iugo** [denodrivers/postgres#386] [Ideas]
A more cautious approach to securing transactions

1. Set a `safe` private property on the `QueryClient`.
2. Change `safe` to `false` every time `begin()` is executed on a connection.
3. Every time `commit()` or `rollback()` is executed on a connection change `safe` to `true`.

What the user has to do:

1. Determine if a connection is secure before starting a new transaction.
2. If it is not safe, you can use `end()` to end the connection and reconnect. You can also execute
    `rollback()` to roll back. Or select `begin()` to commit.

---

更谨慎的 确保事务安全的 办法

1. 在连接上设置一个 `safe` 私有属性.
2. 每次在一个连接上执行 `begin()` 都将 `safe` 改为 `false`.
3. 每次在一个连接上执行 `commit()` 或者 `rollback()` 都将 `safe` 改为 `true`.

使用者要做的:

1. 开始新的事务之前判断一个连接是否安全.
2. 如果不安全, 可以使用 `end()` 结束这个连接, 重新进行连接. 也可以执行
   `rollback()` 进行回滚. 或者选择 `begin()` 进行提交.

**Comment by Soremwar** [comment]
The current approach already accomplishes this, albeit with a little more steps

```ts
if (client.session.current_transaction !== null) {
  await client.createTransaction(transaction_name).rollback();
}
```

I don't recommend ending the connection altogether, since that also removes whatever temporal data was created in the current session

**Reply by iugo** [reply]
`.release()` will `resetSessionMetadata()`.
`resetSessionMetadata()` let `#transaction = null`(session.current_transaction is null).

If the previous client transaction `rollback()` failed, but `release()` succeeded.
It will cause the `client.session.current_transaction` of subsequent reuse of this client to be null.

**Reply by iugo** [reply]
maybe add `.safeRelease()`:

```ts
safeRelease() {
  if (this.session.current_transaction !== null) {
    throw new Error('transaction ongoing');
  }
  this.release();
}
```
